desc/singularities.py

```diff
@@ -0,0 +1,730 @@
+            return put(fj, i, f_temp.squeeze())
+
+        # vmap for inner part found more efficient than fori_loop, especially on gpu,
+        # but for jacobian looped seems to be better and less memory
+        if loop:
+            fj = fori_loop(0, eval_grid.num_nodes, eval_pt_loop, jnp.zeros_like(f))
+        else:
+            fj = vmap(eval_pt_vmap)(jnp.arange(eval_grid.num_nodes))
+
+        f += fj
+        return f, src_data
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    f, _ = fori_loop(0, src_grid.NFP, nfp_loop, (f, src_data))
+
+    return f
+
+
+def _singular_part(
+    eval_data,
+    eval_grid,
+    src_data,
+    src_grid,
+    s,
+    q,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Integrate singular point by interpolating to polar grid."""
+    src_dtheta = src_grid.spacing[:, 1]
+    src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+    eval_theta = jnp.asarray(eval_grid.nodes[:, 1])
+    eval_zeta = jnp.asarray(eval_grid.nodes[:, 2])
+
+    r, w, dr, dw = _get_quadrature_nodes(q)
+
+    h_t = jnp.mean(src_dtheta)
+    h_z = jnp.mean(src_dzeta)
+
+    eta = _chi(r)
+    v = eta * s**2 * h_t * h_z / 4 * abs(r) * dr * dw
+    keys = list(set(["|e_theta x e_zeta|"] + kernel.keys))
+    fsrc = [src_data[key] for key in keys]
+
+    def polar_pt_vmap(i):
+        # evaluate the effect from a single polar node around each singular point
+        # on that singular point. Polar grids from other singularities have no effect
+        dt = s / 2 * h_t * r[i] * jnp.sin(w[i])
+        dz = s / 2 * h_z * r[i] * jnp.cos(w[i])
+        theta_i = eval_theta + dt
+        zeta_i = eval_zeta + dz
+
+        # data interpolated to each eval pt offset by dt,dz
+        src_data_polar = {key: interpolator(val, i) for key, val in zip(keys, fsrc)}
+
+        src_data_polar["zeta"] = zeta_i
+        src_data_polar["theta"] = theta_i
+
+        # eval pts x src pts for 1 polar grid offset
+        # only need diagonal term because polar grid points
+        # don't contribute to other eval pts
+        k = kernel(
+            {key: val for key, val in eval_data.items() if key in keys},
+            src_data_polar,
+            diag=True,
+        )
+        if kernel.ndim == 1:  # so that broadcasting works correctly
+            k = k[:, None]
+        dS = (v[i] * src_data_polar["|e_theta x e_zeta|"])[:, None]
+        fi = k * dS
+        return fi
+
+    def polar_pt_loop(i, f):
+        # this calculates the effect at a single evaluation point, from all others
+        # in a single field period. loop this to get all pts
+        f_temp = polar_pt_vmap(i)
+        return f + f_temp.squeeze()
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    # vmap found more efficient than fori_loop, esp on gpu, but uses more memory
+    if loop:
+        f = fori_loop(0, v.size, polar_pt_loop, f)
+    else:
+        f = vmap(polar_pt_vmap)(jnp.arange(v.size)).sum(axis=0)
+
+    return f
+
+
+def singular_integral(
```

Comment:
seems like the eval data can be in `rpz` but the output quantity will be in `xyz`? should document this if this is the case and possibly add basis argument as well